### ISSUE TITLE
Ability to opt into change detection on `*ReadOnlyItem` in mutable `QueryData` derive

### ISSUE THREAD

**Issue by Themayu** [bevyengine/bevy#12920]
## What problem does this solve or what need does it fill?

Currently, the `QueryData` derive macro transforms fields of the type `&'static mut T` to `Mut<'w, T>` when generating the item type for the query. However, the readonly item type continues to use `&'w T` as the reference type. This makes it impossible to use change detection on the readonly form of mutable queries.

As an example from my code:

```rust
#[derive(QueryData)]
#[query_data(mutable)]
pub struct TextBoxQuery {
    pub data: &'static mut TextBox,
    pub style: &'static mut TextBoxStyle,
    pub text_style: &'static mut TextBoxTextStyle,
}

// derive(QueryData) output
pub struct TextBoxQueryItem<'w> {
    pub data: Mut<'w, TextBox>,
    pub style: Mut<'w, TextBoxStyle>,
    pub text_style: Mut<'w, TextBoxTextStyle>,
}

pub struct TextBoxQueryReadOnlyItem<'w> {
    pub data: &'w TextBox,
    pub style: &'w TextBoxStyle,
    pub text_style: &'w TextBoxTextStyle,
}
```

## What solution would you like?

I would like a way to inform the `QueryData` derive macro to use `bevy_ecs::change_detection::Ref<'w, T>` instead of `&'w T`. I believe it would be useful if we can choose between applying it to the whole struct or to individual fields, like so:

### Example 1: Whole struct application
```rust
#[derive(QueryData)]
#[query_data(mutable, detect_changes)]
pub struct TextBoxQuery {
    pub data: &'static mut TextBox,
    pub style: &'static mut TextBoxStyle,
    pub text_style: &'static mut TextBoxTextStyle,
}

// derive(QueryData) output
pub struct TextBoxQueryItem<'w> {
    pub data: Mut<'w, TextBox>,
    pub style: Mut<'w, TextBoxStyle>,
    pub text_style: Mut<'w, TextBoxTextStyle>,
}

pub struct TextBoxQueryReadOnlyItem<'w> {
    pub data: Ref<'w, TextBox>,
    pub style: Ref<'w, TextBoxStyle>,
    pub text_style: Ref<'w, TextBoxTextStyle>,
}
```

### Example 2: Per field application
```rust
#[derive(QueryData)]
#[query_data(mutable)]
pub struct TextBoxQuery {
    #[query_data(detect_changes)]
    pub data: &'static mut TextBox,

    pub style: &'static mut TextBoxStyle,
    pub text_style: &'static mut TextBoxTextStyle,
}

// derive(QueryData) output
pub struct TextBoxQueryItem<'w> {
    pub data: Mut<'w, TextBox>,
    pub style: Mut<'w, TextBoxStyle>,
    pub text_style: Mut<'w, TextBoxTextStyle>,
}

pub struct TextBoxQueryReadOnlyItem<'w> {
    pub data: Ref<'w, TextBox>,
    pub style: &'w TextBoxStyle,
    pub text_style: &'w TextBoxTextStyle,
}
```

## What alternative(s) have you considered?

The only other option I can see is for me to manually implement `WorldQuery` on `TextBoxQuery`, and I do not wish to manually implement `WorldQuery` due to how brittle it is.

**Comment by Themayu** [comment]
The `detect_changes` option name is up for discussion if necessary. I just chose the most obvious option because I didn't want to spend hours bikeshedding my own feature request before even _posting_ it.

**Comment by alice-i-cecile** [comment]
Seems like a very reasonable request. Mild preference for approach two, with per-field attributes since it's a bit more flexible.

**Comment by Themayu** [comment]
@alice-i-cecile I think you misunderstood. There aren't two separate approaches here - I am proposing supporting _both_ forms of usage.

**Comment by alice-i-cecile** [comment]
Ah, I'm fine with that as well.

**Comment by Themayu** [comment]
I've come up with a (in my opinion) nicer way of achieving the goal here, which has been laid out in #13329.

**Comment by Themayu** [comment]
Closed in favour of #13329.
